File: ruby/A68.rb
# A68
# Ford-Fulkerson法
# Maximum Flow
# 実行時間: 1s以内

class MaxFlow
  Edge = Struct.new(:to, :to_idx, :cap)
  private attr_accessor :graph, :used

  def initialize(size)
    @graph = Array.new(size) { [] }
    @used = [false] * size
  end

  # 頂点from -> 頂点to, 上限cap の辺追加
  def add_edge(from, to, cap)
    from_size = graph[from].size
    to_size = graph[to].size
    graph[from] << Edge.new(to, to_size, cap)
    graph[to] << Edge.new(from, from_size, 0)
  end

  # 最大流量を計算する
  def calc_max_flow(start, goal)
    total_flow = 0
    loop do
      used.fill(false)
      flow = dfs(start, goal)
      return total_flow if flow.zero?
      total_flow += flow
    end
  end

  private
  # goalに辿り着ける流量を1つ求める
  def dfs(pos, goal, max_flow = Float::INFINITY)
    return max_flow if pos == goal
    used[pos] = true
    graph[pos].each do |edge|
      next if used[edge.to] || edge.cap.zero?
      flow = dfs(edge.to, goal, [max_flow, edge.cap].min)
      next if flow.zero?
      edge.cap -= flow
      graph[edge.to][edge.to_idx].cap += flow
      return flow
    end
    0
  end
end

N, M = gets.split.map(&:to_i)
mf = MaxFlow.new(N)
M.times do
  abc = gets.split.map(&:to_i).then { [_1[0].pred, _1[1].pred, _1[2]] }
  mf.add_edge(*abc)
end
puts mf.calc_max_flow(0, N - 1)

File: src/main/kotlin/ktln/A68.kt
/**
 * A68
 * Ford-Fulkerson法
 * Maximum Flow
 * 実行時間: 1s以内
 */

package ktln

import java.io.File
import kotlin.math.min

class MaxFlow(private val size: Int) {

  /**
   * 辺
   *
   * @property to 頂点
   * @property toIdx 頂点のindex
   * @property cap キャパシティ
   */
  data class Edge(val to: Int, val toIdx: Int, var cap: Long)

  private val used = BooleanArray(size + 1) { false }
  private val graph = List(size + 1) { mutableListOf<Edge>() }

  /**
   * 辺追加
   *
   * @param from 頂点
   * @param to 頂点
   * @param cap キャパシティ
   */
  fun addEdge(from: Int, to: Int, cap: Long) {
    val fromIdx = graph[from].size
    val toIdx = graph[to].size
    graph[from].add(Edge(to, toIdx, cap))
    graph[to].add(Edge(from, fromIdx, 0L))
  }

  /**
   * 最大流量計算
   *
   * @param u 頂点
   * @param v 頂点
   * @return 最大流量
   */
  fun calculateMaxFlow(u: Int, v: Int): Long {
    var totalFlow = 0L
    while (true) {
      used.fill(false)
      val flow = dfs(u, v, Long.MAX_VALUE)
      if (flow == 0L) break
      totalFlow += flow
    }
    return totalFlow
  }

  private fun dfs(pos: Int, goal: Int, minFlow: Long): Long {
    if (pos == goal) return minFlow
    used[pos] = true
    graph[pos].forEach { edge ->
      if (edge.cap == 0L || used[edge.to]) return@forEach
      val flow = dfs(edge.to, goal, minOf(minFlow, edge.cap))
      if (flow == 0L) return@forEach
      edge.cap -= flow
      graph[edge.to][edge.toIdx].cap += flow
      return flow
    }
    return 0L
  }
}

fun main() {
  val (N, M) = readLine()!!.split(" ").map(String::toInt)
  val maxFlow = MaxFlow(N)
  repeat(M) {
    val (a, b, c) = readLine()!!.split(" ").let { Triple(it[0].toInt(), it[1].toInt(), it[2].toLong()) }
    maxFlow.addEdge(a, b, c)
  }
  println(maxFlow.calculateMaxFlow(1, N))
}
